builder/dockerfile/instructions/commands.go

```diff
@@ -0,0 +1,381 @@
+package instructions
+
+import (
+	"errors"
+
+	"strings"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/strslice"
+)
+
+// KeyValuePair represent an arbitrary named value (usefull in slice insted of map[string] string to preserve ordering)
+type KeyValuePair struct {
+	Key   string
+	Value string
+}
+
+func (kvp *KeyValuePair) String() string {
+	return kvp.Key + "=" + kvp.Value
+}
+
+// KeyValuePairs is a slice of KeyValuePair
+type KeyValuePairs []KeyValuePair
+
+// BaseCommand is the base class for Dockerfile (String() returns its source code)
+type BaseCommand struct {
+	code string
+	name string
+}
+
+func (c *BaseCommand) String() string {
+	return c.code
+}
+
+// Name of the command
+func (c *BaseCommand) Name() string {
+	return c.name
+}
+
+func newBaseCommand(req parseRequest) BaseCommand {
+	return BaseCommand{code: strings.TrimSpace(req.original), name: req.command}
+}
+
+// SingleWordExpander is a provider for variable expansion where 1 word => 1 output
+type SingleWordExpander func(word string) (string, error)
+
+// SupportsSingleWordExpansion interface marks a command as supporting variable expansion
+type SupportsSingleWordExpansion interface {
+	Expand(expander SingleWordExpander) error
+}
+
+func expandKvp(kvp KeyValuePair, expander SingleWordExpander) (KeyValuePair, error) {
+	key, err := expander(kvp.Key)
+	if err != nil {
+		return KeyValuePair{}, err
+	}
+	value, err := expander(kvp.Value)
+	if err != nil {
+		return KeyValuePair{}, err
+	}
+	return KeyValuePair{Key: key, Value: value}, nil
+}
+func expandKvpsInPlace(kvps KeyValuePairs, expander SingleWordExpander) error {
+	for i, kvp := range kvps {
+		newKvp, err := expandKvp(kvp, expander)
+		if err != nil {
+			return err
+		}
+		kvps[i] = newKvp
+	}
+	return nil
+}
+
+func expandSliceInPlace(values []string, expander SingleWordExpander) error {
+	for i, v := range values {
+		newValue, err := expander(v)
+		if err != nil {
+			return err
+		}
+		values[i] = newValue
+	}
+	return nil
+}
+
+// EnvCommand : ENV key1 value1 [keyN valueN...]
+type EnvCommand struct {
+	BaseCommand
+	Env KeyValuePairs // kvp slice instead of map to preserve ordering
+}
+
+// Expand variables
+func (c *EnvCommand) Expand(expander SingleWordExpander) error {
+	return expandKvpsInPlace(c.Env, expander)
+}
+
+// MaintainerCommand : MAINTAINER maintainer_name
+type MaintainerCommand struct {
+	BaseCommand
+	Maintainer string
+}
+
+// LabelCommand : LABEL some json data describing the image
+//
+// Sets the Label variable foo to bar,
+//
+type LabelCommand struct {
+	BaseCommand
+	Labels KeyValuePairs // kvp slice instead of map to preserve ordering
+}
```

Comment:
I don't agree: the whole point of these PR is to have typed commands. Field names are part of this strongly typed approach, and giving fields generic names just to share the same structure for 2 different aliases seems a bit wrong to me. The extreme case would be to have a common `type Command2TextFields{ Field1, Field2 string}` and having some aliases to this type. But in that case, what is the point of having strongly typed commands if field names have no real meaning ? 

I prefer a bit more code whith more expressiveness and this coherent 1<=> 1 relashionship between typed commands and dockerfile instructions. In longer term it seems more readable / maintainable (what if we want to add a flag on ENV for exemple that has no meaning at all on Label at some point in the future ?)